[PyTorch Numeric Suite] Add module level comparison

hx89 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36701 [PyTorch Numeric Suite] Add module output comparison
* **#36669 [PyTorch Numeric Suite] Add module level comparison**

Add module level comparison API.

Differential Revision: [D21045393](https://our.internmc.facebook.com/intern/diff/D21045393/)

Commit:
[PyTorch Numeric Suite] Add module level comparison

Add module level comparison API.

Differential Revision: [D21045393](https://our.internmc.facebook.com/intern/diff/D21045393/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add module level comparison"


Add module level comparison API.

Differential Revision: [D21045393](https://our.internmc.facebook.com/intern/diff/D21045393/)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_numeric_suite.py:
```diff
@@ -22,3 +26,49 @@ def test_compare_weights(self):
         self.assertEqual(len(weight_dict), 1)
         for k, v in weight_dict.items():
             self.assertTrue(v["float"].shape == v["quantized"].shape)
+
```
Will be good to write this as a loop to reduce code:
``` python
 def test_compare_weights(self):
      model_list = [AnnotatedConvModel(), AnnotatedConvBnReLUModel()]
      for model in model_list:
             model.eval()
             ...
```

raghuramank100 commented on test/quantization/test_numeric_suite.py:
```diff
@@ -22,3 +26,49 @@ def test_compare_weights(self):
+            annotated_conv_model, default_eval_fn, self.img_data
+        )
+        data = self.img_data[0][0]
+        module_swap_list = [torch.nn.Conv2d]
```
Similar comments as above. If you make module_swap_list = [ torch.nn.Conv2d, nn.intrinsic.modules.fused.ConvReLU2d] , it should work.

raghuramank100 commented on test/quantization/test_numeric_suite.py:
```diff
@@ -22,3 +26,49 @@ def test_compare_weights(self):
+            annotated_conv_bn_relu_model, default_eval_fn, self.img_data
+        )
+        module_swap_list = [nn.intrinsic.modules.fused.ConvReLU2d]
+        ob_dict = compare_model_stub(
```
Please add an additional test for swapping a user defined module here. For example
```python
class ModelWithSubModules(torch.nn.Module):
    def __init__(self):
        super(ModelWithSubModules, self).__init__()
        self.qconfig = default_qconfig
        self.mod1 = subModule()
        self.conv = torch.nn.Conv2d(...)
        self.quant = QuantStub()
        self.dequant = DeQuantStub()

def forward(self, x):
     x = self.quant(x)
     x = self.mod1(x)
     x = self.conv(x)
     x= self.dequant(x)
    return x
```
 The test should add a stub only to mod1

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -39,3 +42,139 @@ def compare_weights(float_dict, quantized_dict):
+        return prefix if prefix == "" else prefix + "."
+
+    if hasattr(mod, observer_name):
+        if observer_name == "logger":
```
Instead of matching strings, suggest having a baseclass for logger and for tensor observers. As long as someone inherits this base-class, we can match.
i.e do
 ```
 if type(observer) == logger:
  ```

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -39,3 +42,139 @@ def compare_weights(float_dict, quantized_dict):
     return weight_dict
+
+
+def _get_observer_dict(mod, target_dict, observer_name, prefix=""):
```
Suggest making this a top level API, i.e remove the _ Also, instead of observer name, lets go with observer type.

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -39,3 +42,139 @@ def compare_weights(float_dict, quantized_dict):
+        _get_observer_dict(child, target_dict, observer_name, module_prefix)
+
+
+class Logger(nn.Module):
```
Create a base class that only has a stats dict but no forward method and inherit logger from that class.

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -39,3 +42,139 @@ def compare_weights(float_dict, quantized_dict):
+    modules to do the comparison.
+    """
+
+    def __init__(self, q_module, float_module, Logger):
```
We need to add more documentation to explain the arguments for the init function for the class.

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -39,3 +42,139 @@ def compare_weights(float_dict, quantized_dict):
+        return output
+
+
+def _prepare_model_with_stubs(float_module, q_module, module_swap_list, Logger):
```
Lets make this a top level function, the reason is that users can insert arbitrary code between preparing and collecting stats. Perhaps they want to run several batches of data etc

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -39,3 +42,139 @@ def compare_weights(float_dict, quantized_dict):
+        q_module._modules[key] = value
+
+
+def compare_model_stub(float_model, q_model, module_swap_list, data, Logger=Logger):
```
Wondering if we need this as a top level function or keep this as an internal/test only function and move it to the test code.

hx89 replied:
Maybe we can keep it as top level function since it's easy for user to use as a one-liner to quickly check the model accuracy?

raghuramank100 replied:
Sure, as long as we have the building blocks as user facing functions, we are ok.

raghuramank100 commented on test/quantization/test_numeric_suite.py:
```diff
@@ -22,3 +26,49 @@ def test_compare_weights(self):
+            data,
+        )
+        self.assertEqual(len(ob_dict), 1)
+        for k, v in ob_dict.items():
```
Also, please add a test for functional modules (add and concat) to ensure that the API works for that also.

raghuramank100 reviewed (changes requested):
Great start!.  Few suggestions for further improvement

Commit:
Update on "[PyTorch Numeric Suite] Add module level comparison"


Add module level comparison API.

Differential Revision: [D21045393](https://our.internmc.facebook.com/intern/diff/D21045393/)

[ghstack-poisoned]

hx89 commented:
Address comments.

Commit:
Update on "[PyTorch Numeric Suite] Add module level comparison"


Add module level comparison API.

Differential Revision: [D21045393](https://our.internmc.facebook.com/intern/diff/D21045393/)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -39,3 +43,203 @@ def compare_weights(float_dict, quantized_dict):
+        self.stats["quantized"] = None
+
+    def forward(self, x, y):
+        print("in RecordingLogger forward")
```
Please remove print

raghuramank100 commented on test/quantization/test_numeric_suite.py:
```diff
@@ -1,24 +1,155 @@
+        model = ModelWithFunctionals().eval()
+        q_model = copy.deepcopy(model)
+        q_model.qconfig = torch.quantization.get_default_qconfig("fbgemm")
+        q_model = prepare(q_model)
```
You can remove the deepcopy by calling prepare with inplace=False

raghuramank100 reviewed (approved):
Looks good, have a few comments

Commit:
Update on "[PyTorch Numeric Suite] Add module level comparison"


Add module level comparison API.

Differential Revision: [D21045393](https://our.internmc.facebook.com/intern/diff/D21045393/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add module level comparison"


Add module level comparison API.

Differential Revision: [D21045393](https://our.internmc.facebook.com/intern/diff/D21045393/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add module level comparison"


Add module level comparison API.

Differential Revision: [D21045393](https://our.internmc.facebook.com/intern/diff/D21045393/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Numeric Suite] Add module level comparison"


Add module level comparison API.

Differential Revision: [D21045393](https://our.internmc.facebook.com/intern/diff/D21045393/)

[ghstack-poisoned]
